### PR TITLE
Linked proofs verifier

### PR DESCRIPTION
**Pull request by Kolezhniuk**

### CODE REVIEW

File: contracts/verifiers/UniversalVerifier.sol [hunk 1]
```diff
@@ -317,4 +320,31 @@ contract UniversalVerifier is OwnableUpgradeable, IUniversalVerifier {
     ) public view returns (uint256) {
         return _getMainStorage().proofs[user][requestId].storageFields[key];
     }
+
+    /// @notice Gets the list of reuest IDs and verifies the proofs are linked
+    /// @param requestIds the list of reuest IDs
+    /// @return Whether the proofs are linked
+    function verifyLinkedProofs(uint64[] calldata requestIds) public view returns (bool) {
+        Proof storage firstProof = _getMainStorage().proofs[_msgSender()][requestIds[0]];
```
Comment (vmidyllic) [inline contracts/verifiers/UniversalVerifier.sol:328]: check that len of requests if more than 1

File: contracts/verifiers/UniversalVerifier.sol [hunk 2]
```diff
@@ -317,4 +320,31 @@ contract UniversalVerifier is OwnableUpgradeable, IUniversalVerifier {
     ) public view returns (uint256) {
         return _getMainStorage().proofs[user][requestId].storageFields[key];
     }
+
+    /// @notice Gets the list of reuest IDs and verifies the proofs are linked
+    /// @param requestIds the list of reuest IDs
```
Comment (vmidyllic) [inline contracts/verifiers/UniversalVerifier.sol:325]: request

File: contracts/verifiers/UniversalVerifier.sol [hunk 3]
```diff
@@ -317,4 +320,31 @@ contract UniversalVerifier is OwnableUpgradeable, IUniversalVerifier {
...
+        for (uint256 i = 1; i < requestIds.length; i++) {
+            Proof storage proof = _getMainStorage().proofs[_msgSender()][requestIds[i]];
+            uint256 linkID = proof.storageFields[LINKED_PROOF_KEY];
+
+            if (linkID == 0) {
+                revert("linkID cannot be 0.");
+            }
+
+            if (linkIdToCompare != linkID) {
+                revert("linkID verification failed");
```
Comment (vmidyllic) [inline contracts/verifiers/UniversalVerifier.sol:344]: we need to find a way to specify specific requests have specific link id but they are not linked

**Review by vmidyllic — DISMISSED** [review]

File: test/verifier/universal-verifier.v3.test.ts [hunk 4]
```diff
@@ -101,4 +100,32 @@ describe("Universal Verifier V3 validator", function () {
 
     // TODO make some test with correct UserID but with wrong challenge
   });
+
+  it("Test linked proofs", async () => {
```
Comment (AndriianChestnykh) [inline test/verifier/universal-verifier.v3.test.ts:104]: May be move this test to a separate file?

Comment (AndriianChestnykh) [reply test/verifier/universal-verifier.v3.test.ts:104]: yes, the tests are already separated for the UV itself, there are V2, V3 and events tests. This is just to keep it consistent.

File: contracts/verifiers/UniversalVerifier.sol [hunk 5]
```diff
@@ -317,4 +329,37 @@ contract UniversalVerifier is Ownable2StepUpgradeable, IUniversalVerifier {
...
         return _getMainStorage().proofs[user][requestId].storageFields[key];
     }
+
+    /// @notice Gets the list of request IDs and verifies the proofs are linked
+    /// @param requestIds the list of request IDs
+    /// Throws if the proofs are not linked
+    function verifyLinkedProofs(uint64[] calldata requestIds) public view {
+        require(requestIds.length > 1, "Linked proof verification needs more than 1 request");
+
+        Proof storage firstProof = _getMainStorage().proofs[_msgSender()][requestIds[0]];
```
[the remark this answers is not in the document]
Comment (AndriianChestnykh) [reply contracts/verifiers/UniversalVerifier.sol]: For `submitZKPResponse` and `setZKPRequest` it is important from security standpoint, as they do transactions. `verifyLinkedProof` is just a view informational function. There is no harm to add this flexibility.

[the remark this answers is not in the document]
Comment (vmidyllic) [reply contracts/verifiers/UniversalVerifier.sol]: parameter is ok and right

**Review by vmidyllic — APPROVED** [review]

File: test/verifier/universal-verifier-linked-proofs.test.ts [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    expect(await verifier.verifyLinkedProofs(signerAddress, [100, 103])).not.to.throw;
+  });
+
+  it("should linked proof validation fail", async () => {
+    await expect(verifier.verifyLinkedProofs(signerAddress, [100, 101])).to.be.revertedWith(
+      "LinkedProofError"
+    );
+    await expect(verifier.verifyLinkedProofs(signerAddress, [102, 103])).to.be.revertedWith(
+      "LinkedProofError"
+    );
```
Comment (AndriianChestnykh) [inline test/verifier/universal-verifier-linked-proofs.test.ts:59]: Any chance to check the other error params?

**Review by AndriianChestnykh — APPROVED** [review]

**Pull request merged**